### PR TITLE
Add emulation data preprocessing transforms to fv3fit

### PR DESCRIPTION
**Pull request by frodre**
This PR adds configurable data pre-processing elements to fv3fit, which are generally useful for translating from the batched xarray datasets to arrays/tensors used in ML model training.  This PR is the first in a series that will be standardizing some tools for our emulation projects.  

Added public API:
- `fv3fit.emulation.data`  data transforms and configurable pipeline entrypoints
    - `TransformConfig` for a configurable standard input transformation pipeline for keras
    - `get_nc_files` to grab a list of local or remote netcdf files in a directory
    - `nc_files_to_tf_dataset` convert a list of netcdf files to a `tf.data.Dataset`
    - `nc_dir_to_tf_dataset` convert a directory of netcdf files to a `tf.data.Dataset`
    - `batches_to_tf_dataset` a convenience function to go from batched data from out of `loaders` to a `tf.data.Dataset`
    - `transforms` provides access to all transforms, which is useful for things like notebook prototyping

- [x] Tests added

### CODE REVIEW

**Comment by frodre** [comment]
[ ] - Try out dumpable transform class for legacy
[ ] - Add tests for transforms
[ ] - finish config -> pipeline re-integration

File: external/fv3fit/fv3fit/emulation/data/transforms.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    return {
+        key: tf.convert_to_tensor(data, dtype=dtype) for key, data in dataset.items()
+    }
+
+
+@curry
+def select_antarctic(dataset: xr.Dataset, sample_dim_name="sample") -> xr.Dataset:
+    """
+    Select only points below 60 S.  Requires 'latitude' in dataset and expects
+    units in radians
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/data/transforms.py:44]: I believe the default for lats/lons is radians in fv3, but let me know if this is not the case.

File: external/fv3fit/fv3fit/emulation/data/config.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    ----
+        input_variables: Variables to include as inputs for training
+        output_variables: Variables to include as targets for training
+        antarctic_only: Limit data to < 60 S.  Requires latitude exists
+            as a field in the dataset
+        use_tensors: Converts data to float32 tensors instead of numpy arrays
+            vertical_subselection: Limit the feature dimension of a variable
+            to a specified range. Loaded in as slices from a 2 or 3 item
+            sequence.
+        transforms: Sequence of extra transform configurations to combine
```
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/data/config.py:185]: This is inherited from the TransformConfig class, but I am open to removing it since there's no immediate use case right now, and the fact that this configuration is not meant for a large amount of flexibility.

**Review by AnnaKwa — CHANGES_REQUESTED** [review]
Thanks Andre, looks good! There are a couple minor requests in the comments. 

My main less-specific question/concern was that there looks to be two main types of transforms (that work on either xarray datasets or ndarray/tensor data) which need to be on different sides of the to_array/tensor transform. If the only xarray-input transform that will ever be implemented is the Antarctica filter I suppose that is fine, but if more get added in the future I think some distinction between these types of transforms would be useful, where xarray-input ones must occur first if there is a conversion to ndarray/tensors.

Could you also start an emulation section in the fv3fit docs and if you don't plan to add a vcm-workflow-control example for the emulator could you start to fill in the relevant bits as you merge this series of PRs?

**Comment by nbren12** [comment]
Related to #1265 .

File: external/fv3fit/tests/emulation/data_transform_mock.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+from toolz.functoolz import curry
```
Comment (nbren12) [inline external/fv3fit/tests/emulation/data_transform_mock.py:1]: It seems like these are only used in `test_data_config`. Can you move them to that model?

File: external/fv3fit/tests/emulation/test_data_transforms.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
+import pytest
+import numpy as np
+import tensorflow as tf
+import xarray as xr
+from fv3fit.emulation.data import transforms as xfm
```
Comment (nbren12) [inline external/fv3fit/tests/emulation/test_data_transforms.py:5]: nit: Some style guides (e.g. Google) disallow this kind of nonstandard renaming and I don't think `transforms` is an overly long name to use below. Is there some reason we need a three letter abbreviation for this?

```suggestion
from fv3fit.emulation.data import transforms
```

**Review by nbren12 — CHANGES_REQUESTED** [review]
The actual transform functions look useful and well documented, but I think the configuration front-end is a bit heavyweight for what it does and likely to change in the future. To me, a static configuration should provide support for the standard cases, and if something more complicated needs to be done the source code should be modified rather than making the config file act like source code. It is easier to test python code than yaml code.

I suggest
- removing `TransformConfigItem` since we probably don't want our yaml files to turn into python scripts.
- Refactor `TransformConfig` so that it is a callable object that can be used like this
```
transform = TransformConfig(only_antarctica=True)
transform(data)
```
This would simplify the implementation of `get_pipeline_config` so that it instead does the transforms, rather than building up a list of transforms to be done. That kind of lazy scheduling often makes it harder to debug these kinds of transformations easily.

File: external/fv3fit/fv3fit/emulation/data/config.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+    name: str
+    args: Union[Sequence[Any], Mapping[str, Any]] = dataclasses.field(
+        default_factory=list
+    )
+    kwargs: Mapping[str, Any] = dataclasses.field(default_factory=dict)
+
+    def __post_init__(self):
+        if isinstance(self.args, Mapping):
+            self.args = list(self.args.values())
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/config.py:92]: See comment about removal.  I originally had this in there because I was only using mappings (easier to read from yaml) for argument input into the functions.  I think I ever bring config specified custom pipelines, I'll make sure it can just use all keyword arguments instead.

File: external/fv3fit/fv3fit/emulation/data/transforms.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+import tensorflow as tf
+import xarray as xr
+from toolz.functoolz import curry
+from typing import Mapping, Sequence, Tuple, Union
+
+
+logger = logging.getLogger(__name__)
+
+NumericContainer = Union[np.ndarray, xr.DataArray, tf.Tensor]
+StandardizeInfo = Mapping[str, Tuple[NumericContainer, NumericContainer]]
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/transforms.py:17]: This was a leftover from when I had standardization in the pipeline.

File: external/fv3fit/fv3fit/emulation/data/transforms.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    logger.debug(f"input vars: {input_variables}")
+    logger.debug(f"output vars: {output_variables}")
+    inputs_ = tuple([dataset[key] for key in input_variables])
+    outputs_ = tuple([dataset[key] for key in output_variables])
+
+    return inputs_, outputs_
+
+
+@curry
+def maybe_subselect(
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/transforms.py:74]: I adjusted to `maybe_subselect_feature_dim` and `expand_single_dim_data`.

File: external/fv3fit/fv3fit/emulation/data/config.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    output_variables: Sequence[str] = dataclasses.field(default_factory=list)
+    antarctic_only: bool = False
+    use_tensors: bool = True
+    vertical_subselections: Union[Mapping[str, slice], None] = None
+
+    def get_transform_pipeline(self):
+
+        transform_funcs = []
+
+        if self.antarctic_only:
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/config.py:223]: Antarctic only reduces samples to only the antarctic region.  If false, it will not do any reduction (i.e., full dataset).

File: external/fv3fit/fv3fit/emulation/data/transforms.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    subselection_map: Mapping[str, slice], dataset: Union[ArrayDataset, TensorDataset]
+) -> Union[ArrayDataset, TensorDataset]:
+    """
+    Subselect from the feature dimension if specified in the map.
+    """
+    new_ds = {}
+    for vname, arr in dataset.items():
+        if vname in subselection_map:
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/transforms.py:82]: I was hesitant to put this in since I often do changes to inputs and output fields, but I had a pretty static set of vertical limits in place on certain types of fields (e.g., moisture).  I'm open to putting it in if you think allowing over specification in the config yaml is no good.

File: external/fv3fit/fv3fit/emulation/data/transforms.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+def to_tensors(dataset: InputDataset, dtype: tf.DType = tf.float32) -> TensorDataset:
+    """Convert a dataset to tensors with specified dtype."""
+    logger.debug("Converting dataset to tensor dataset")
+    return {
+        key: tf.convert_to_tensor(data, dtype=dtype) for key, data in dataset.items()
+    }
+
+
+@curry
+def select_antarctic(dataset: xr.Dataset, sample_dim_name="sample") -> xr.Dataset:
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/transforms.py:41]: Good point!  I couldn't think of an easy way to ensure the I/O type consistency of the pipeline is preserved or checked.  This is another reason why I removed the custom pipeline thru yaml from the PR.  Intentional standard pipelines in the code at least enforce that the pipeline order makes sense.

File: external/fv3fit/fv3fit/emulation/data/config.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+TransformItemSpec = Union[Mapping, _TransformConfigItem]
+TransformConfigSpec = Dict[str, Sequence[TransformItemSpec]]
+
+
+@dataclasses.dataclass
+class TransformConfig:
+    """
+    Specify a custom transform pipeline for data
+
+    Args
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/config.py:131]: If I combine all the docstring styles it gives the user more power. *thanos_gif*

File: external/fv3fit/fv3fit/emulation/data/config.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+
+        input_transforms = InputTransformConfig.from_dict(config)
+        pipeline = input_transforms.get_transform_pipeline()
+    """
+
+    input_variables: Sequence[str] = dataclasses.field(default_factory=list)
+    output_variables: Sequence[str] = dataclasses.field(default_factory=list)
+    antarctic_only: bool = False
+    use_tensors: bool = True
+    vertical_subselections: Union[Mapping[str, slice], None] = None
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/config.py:217]: I updated before I realized this was a suggestion

File: external/fv3fit/fv3fit/emulation/data/config.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+        with open("config.yaml", "r") as f:
+            config = yaml.load(f, Loader=SliceLoader)
+
+        input_transforms = InputTransformConfig.from_dict(config)
+        pipeline = input_transforms.get_transform_pipeline()
+    """
+
+    input_variables: Sequence[str] = dataclasses.field(default_factory=list)
+    output_variables: Sequence[str] = dataclasses.field(default_factory=list)
+    antarctic_only: bool = False
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/config.py:215]: Yeah, I was thinking similarly.

File: external/fv3fit/fv3fit/emulation/data/config.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+from toolz.functoolz import compose_left
+from typing import Any, Callable, Dict, Mapping, Sequence, Tuple, Union
+
+from . import transforms
+
+
+logger = logging.getLogger(__name__)
+
+
+class SliceLoader(yaml.SafeLoader):
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/config.py:17]: Removed

File: external/fv3fit/tests/emulation/data_transform_mock.py (hunk 3 above)
Comment (frodre) [reply external/fv3fit/tests/emulation/data_transform_mock.py:1]: Removed with the custom pipeline config removal.  I was creating the module to mock the other module which also has mostly transforms available in the namespace for `getattr`.

File: external/fv3fit/tests/emulation/test_data_transforms.py (hunk 4 above)
Comment (frodre) [reply external/fv3fit/tests/emulation/test_data_transforms.py:5]: Same story as above suggestion.

**Comment by frodre** [comment]
I decided to remove the custom transform pipeline since it was just some personal use cases I had in mind, but I generally agree that the flexibility probably adds too much complexity and potentially confusing aspects.

I added in the basics for loading from a directory of netcdfs.  @AnnaKwa I'll make sure to add an example of usage into vcm-workflow-control, but I think I'll wait on docs entry until this has stabilized.

**Review by AnnaKwa — APPROVED** [review]
Thanks Andre! The removal of the custom pipeline and unused transform arg resolves the other part of my comments. The docs example is nice and concise. Could you add a line or two explaining the allowed data format for the input data into the pipeline (e.g. stacked into sample dim)?

File: external/fv3fit/fv3fit/emulation/data/load.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        path: Local or remote gcs path to netCDF directory
+    """
+
+    fs = get_fs(path)
+    files = list(fs.glob(os.path.join(path, "*.nc")))
+
+    return files
+
+
+def batched_to_tf_dataset(
```
Comment (nbren12) [inline external/fv3fit/fv3fit/emulation/data/load.py:26]: This function does a few too many things. It transforms a sequence into a tf Dataset, than maps over this dataset, and sets a pretuning strategy. It feels like this function should just be copied verbatim to wherever it is used since it is very abstract and in all the common cases the input data are actually a list of file names? Does this not work?
```
tf.data.Dataset.from_tensor_slices(filenames).flat_map(transform).prefetch(tf.data.AUTOTUNE)
```

The name is also quite specific name given how abstract the function is. This function will work with a very broad class of inputs, not just "batched_sources". Also it mostly is a mapping operation so should have the word "map".

File: external/fv3fit/fv3fit/emulation/data/transforms.py [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+logger = logging.getLogger(__name__)
+
+NumericContainer = Union[np.ndarray, xr.DataArray, tf.Tensor]
+ArrayDataset = Mapping[str, np.ndarray]
+TensorDataset = Mapping[str, tf.Tensor]
+AnyDataset = Mapping[str, NumericContainer]
+
+
+def open_netcdf_dataset(path: str) -> xr.Dataset:
```
Comment (nbren12) [inline external/fv3fit/fv3fit/emulation/data/transforms.py:24]: This I/O code is dissimilar to the other contents of this module. I suggest moving it someplace else.

File: external/fv3fit/tests/emulation/test_data_transforms.py [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+@pytest.mark.parametrize(
+    "lats, data, expected",
+    [
+        ([-55, -60, -65], [1, 2, 3], [3]),
+        ([55, 60, 65], [1, 2, 3], []),
+        ([-61, -65, -80], [1, 2, 3], [1, 2, 3]),
+    ],
+)
+def test_select_antarcic(lats, data, expected):
```
Comment (nbren12) [inline external/fv3fit/tests/emulation/test_data_transforms.py:67]:
```suggestion
def test_select_antarctic(lats, data, expected):
```

**Review by nbren12 — CHANGES_REQUESTED** [review]
Thanks Andre! This is looking almost ready and much more useful after adding the netcdf capability.

I suggest deleting the `batched_to_tf_dataset` function (see below) and either replacing it with a high level helper function:
```
open_netcdf_folder(url: str, config: TrainingConfig)
```
or show my suggestion below in the docs instead.

File: external/fv3fit/fv3fit/emulation/data/load.py (hunk 15 above)
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/load.py:26]: > It feels like this function should just be copied verbatim to wherever it is used since it is very abstract and in all the common cases the input data are actually a list of file names? Does this not work?

Yeah, it doesn't work out with the one-liner since the `transform` is not a tensorflow-native operation and it tries to generate graph ops for everything in the provided transform.  I tried hiding the transform in a sequence-like structure using `loaders.batches._sequences.Map`, but that didn't work either.  We could wrap transform in a `tf.py_function`, but that doesn't handle nested structure outputs (i.e., two groups of tuples) natively from what I could tell.  I settled on using the `from_generator` because it seemed the most straightforward to understand.

I made this function a bit more flexible because I was thinking there still might be cases where I use the sequences created by loaders (e.g., non-subsampled data) instead of an input list of netCDF files.

I also had the one-liner convenience function going from files to a tf dataset, but it felt like something I'd put in the place where the training workflow is run.  Also, it wasn't clear what level of convenience I should provide for the transform portion, such as the option to initialize from a config file or a dict if provided.  Would be fine throwing this back in somewhere if you'd find it useful.

>This function will work with a very broad class of inputs, not just "batched_sources". Also it mostly is a mapping operation so should have the word "map".

Maybe `sequence_map_to_tf_dataset`?

> This function does a few too many things.

I can let the user decide whether or not they want to use the pre-fetch operation.  Since we can't generally parallelize when using a generator or py_funcs with map, I figured adding this by default would at least give users the performance boost for tensorflow training.

**Review by nbren12 — APPROVED** [review]
Approved following a dicussion with Andre. We agreed to make the controversial `batched_to_tf_dataset` a private function to be used by two public APIs:
1. open folder of netcdf with options
2. open list of netcdf filenames with options.

File: external/fv3fit/docs/emulation_data.rst [hunk 18]
```diff
@@ -0,0 +1,34 @@
+.. emulation_data_:
+
+Emulation Data Tools
+====================
+
+``fv3fit.emulation.data`` contains a set of composable transforms useful for building training input pipelines, an input transform configuration, and convenience functions for creating tensorflow datasets from batched data.
+
```
[the remark this answers is not in the document]
Comment (frodre) [reply external/fv3fit/docs/emulation_data.rst:7]: I added a description note to the example, and then adjusted the `TransformConfig` docstring to be explicit about the data it expects.

File: external/fv3fit/fv3fit/emulation/data/transforms.py (hunk 16 above)
Comment (frodre) [reply external/fv3fit/fv3fit/emulation/data/transforms.py:24]: Moved to `fv3fit.emulation.data.io`
